src/ja/server/scheduler/default_algorithm.py

```diff
@@ -17,16 +31,150 @@ def __init__(self,
         """!
         Initialize the scheduling algorithm.
 
-        @param cost_function The cost function to use.
-        @param non_preemptive_distribution_policy The policy to use when distributing jobs which cannot preempt other
           jobs. The list of jobs to be preempted returned by JobDistributionPolicy.assign_machine is ignored.
         @param blocking_distribution_policy The policy to use when choosing a machine to reserve for a blocking job.
           The list of jobs to be preempted returned by JobDistributionPolicy.assign_machine is ignored.
         @param preemptive_distribution_policy The policy to use when choosing a machine to reserve for a job which can
           preempt other jobs.
         """
+        self._cost_func = cost_function
+        self._non_preemptive_policy = non_preemptive_distribution_policy
+        self._blocking_policy = blocking_distribution_policy
+        self._preemptive_policy = preemptive_distribution_policy
+        self._reserved_machines: Dict[str, str] = {}  # Job UID -> Machine UID
+        self._cost_cache: Dict[str, float] = {}  # Job UID -> Effective cost
+
+    def _set_state(self,
+                   job: Job,
+                   schedule: ServerDatabase.JobDistribution,
+                   machine: WorkMachine,
+                   new_status: JobStatus) -> None:
+        if job.status is JobStatus.RUNNING or job.status is JobStatus.PAUSED:
+            machine.resources.deallocate(get_allocation_for_job(job))
+
+        job.status = new_status
+        machine.resources.allocate(get_allocation_for_job(job))
```

Comment:
This function only works for some status changes since it wouldn't work with crashed or cancelled as a new status right?

Reply:
Yes, crashed or cancelled is not handled by the scheduling algorithm.